### ISSUE TITLE
Intersection memory

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#475]
When I run Haowen's decode.py using a previously prepared LG.pt, it fails with out-of-memory in decoding.
The issue is likely that intersect_dense_pruned uses a lookup table whose size is (num-states in LG t* minibatch size).
It would be better to either replace this with some kind of hash table or rework the algorithm completely somehow.
Probably the better solution is to develop some  kind of hash-table-like structure in k2.  Just leaving the issue here for now.
```
  File "decode.py", line 155, in main
    results = decode(dataloader=test_dl,
  File "decode.py", line 60, in decode
    lattices = k2.intersect_dense_pruned(LG, dense_fsa_vec, 2000.0, 20.0,
  File "/ceph-dan/.local/lib/python3.8/site-packages/k2/autograd.py", line 445, in intersect_dense_pruned
    _IntersectDensePrunedFunction.apply(a_fsas, b_fsas, out_fsa, search_beam,
  File "/ceph-dan/.local/lib/python3.8/site-packages/k2/autograd.py", line 145, in forward
    ragged_arc, arc_map_a, arc_map_b = _k2.intersect_dense_pruned(
RuntimeError: CUDA out of memory. Tried to allocate 3.04 GiB (GPU 0; 31.75 GiB total capacity; 26.80 GiB already allocated; 2.56 GiB free; 27.86 GiB reserved in total by PyTorch)
```

**Comment by arunkumaras10** [comment]
Hi @danpovey. Any update on this issue? I am also facing CUDA out of memory issue especially when using OnlineDenseIntersector with a precompiled HLG. Need to use a larger mini batch size while decode. Any tips or hacks around this issue would be greatly appreciated. Thanks in advance.

**Comment by pkufool** [comment]
@arunkumaras10  I think your problem is different from this issue. Do you have long audio files? BTW, you can open a new issue to describe your problem.

**Comment by danpovey** [comment]
yes, sizes are of interest, e.g. what frame are we on, what input dimensions, what batch size, what HLG size, etc.
